### ISSUE TITLE
cache: a different image of the same size reuses the old image's state - the prefix cache compares IMAGE_PAD ids only, so 7 of 12 colour-probe requests were answered for the first image (prefill 0 of 1047)

### ISSUE THREAD

**Issue by nibor1896** [nibor1896/crow-nest#114]
## Symptom

On 2026-09-24 10:26-10:27 UTC, `tools/vit-colorprobe.py` against robin's `serve` (crow-nest @ 0da7420, F16 projector, #108) sent one 256x256 solid-colour PNG plus the same question per request. The model answered for the FIRST image on every later same-size request: red->Red, green->Red, blue->Red, white->Red, ...; a horizontal bar was read as "Vertical"; 4 of 11 answers right (probe run by the lead session, answers as reported there). A different-size image (grid (1,48,86)) went COLD and was answered correctly.

Impact: any repeated same-size image at the same prompt position is answered from a stale image. Crow's judge (nibor1896/crow#266) sends rubric + one image per call, so it may score the previous image.

## Evidence

`~/.local/state/crow/logs/engine.log` (UTC), 12 `[cache]` decisions between 10:26:45 and 10:27:08:

```
10:26:45.871 [vit-cache] 1 image(s): 1 through the tower, 0 cached; cache 1 entries
10:26:45.873 [cache] COLD L 0 (held 0), P 0, ... prefill 1047 of 1047 tok
10:26:48.254 [vit-cache] 1 image(s): 1 through the tower, 0 cached; cache 2 entries
10:26:48.266 [cache] WARM L 1047 (held 1048), P 1047, ... prefill 0 of 1047 tok
... (the same pair 5 more times, cache 3..7 entries, every image a tower MISS = a new image)
10:27:00.115 [cache] WARM L 1049 (held 1050), P 1049, ... prefill 0 of 1049 tok
```

- 7 of 12 decisions are `WARM ... prefill 0 of N`: the tower ran on a NEW image (`1 through the tower, 0 cached`), yet no row of the prompt was prefilled, so the new visual embeddings were never spliced.
- Cause in code @ 0da7420: `vit::expand_ids` (`engine/src/vit.rs:1311`) expands each image to `n_visual` copies of `IMAGE_PAD` = 248056 (`vit.rs:46`). The prefix cache compares ids only: `PrefixCache::decide_for` -> `common_prefix_len(history, ids)` (`engine/src/cache.rs:431`, `:435`, `:214`), called from `chat_generate` (`engine/src/bin/serve.rs:3970`). Two different images of the same grid are the same 1024-id run, so `L` = the whole prompt and the #100 zero-prefill path (`serve.rs:4013`, `restore_logits`) serves the OLD image's logits row. The splice itself (`gen.rs:3811`, `plan.map[prompt_row]`) only runs for prefilled rows.
- The image's content hash already exists: `Vit::build_plan` keys the tower-output LRU on a 64-bit hash of the image bytes (`vit.rs:1466`); it never reaches the prefix cache.

## Attribution

MEASURED FAILURE, this repo (`cache.rs` / `serve.rs`). Not the tower: every image went through it (`1 through the tower`), and #73's in-flight-D2H bug is fixed (`Vit::run` syncs). Not the model: the COLD request with a different grid was answered right. #100 made it visible at the request length (zero prefill); before #100 the same `L` would have rolled back to the snapshot one below and still skipped the image rows. Related, not duplicates: #100 / #101 (slot bookkeeping, text only), #66 / #73 (image path).

## Research

- llama.cpp server (local source `~/.local/share/crow/src/llama.cpp` @ cbca449): `server_tokens` stores `LLAMA_TOKEN_NULL` for every media token and keeps `map_idx_to_media[start] = chunk` (`tools/server/server-common.cpp:488-507`, `push_back(const mtmd_input_chunk*)`). `server_tokens::get_common_prefix` (`server-common.cpp:680-722`) compares text tokens by id and, at a media position, compares `mtmd_input_chunk_get_id` (the image id) AND `mtmd_input_chunk_get_n_tokens`; if either differs it returns the chunk's first index. The id is `hash_sha256_hex(buf, len)` of the encoded file bytes, "use sha256 to prevent cache poisoning" (`tools/mtmd/mtmd-helper.cpp:378-387`). `keep_first(n)` refuses to cut inside an image and drops media entries at or above `n` (`server-common.cpp:636-667`).
- vLLM automatic prefix caching (https://docs.vllm.ai/en/latest/design/prefix_caching.html, "A hashing example with multi-modality inputs"): block hashes include "Extra hashes: ... multi-modality input hashes", because placeholder tokens alone cannot tell images apart: "we encode the image hash generated by the frontend image processor."
- Both: equality of the id stream is not enough; the image identity must be part of the comparison at the image's position. Read only, not measured locally.

## Proposed fix

Checked against 0da7420 (main).

A. **Image spans next to the ids (chosen, the llama.cpp rule).** `VisionPlan` carries `spans: Vec<ImageSpan { start, len, hash }>` (hash = the same `image_key(bytes)` the tower LRU uses). `Engine` holds `history_images` beside `history`. New pure `cache::common_prefix_len_mm(a, a_imgs, b, b_imgs)`: `L` = id prefix, cut to the `start` of the first image span on either side that has no identical span on the other. `decide_mm` uses it; `serve` passes the plan's spans and sets `history_images` after the prefill. Every path that rewrites `history` keeps the spans consistent: `rollback` truncates (spans starting at or above `P` dropped), `reset_to_zero` clears, slot-file restore (`slot.rs`) clears (a file carries ids, not image identity, so a restored image never matches).
```rust
// before (cache.rs:435)            // after
let l = common_prefix_len(h, ids);  let l = common_prefix_len_mm(h, held_imgs, ids, imgs);
```
B. Replace each `IMAGE_PAD` in a comparison copy of the ids by a hash-derived pseudo id. Same effect, but needs a second id vector of history length kept in step with `history` on every push/truncate; A only touches the few image records.
C. Always COLD when a request has images. Correct but throws away the WARM reuse agent loops need (Crow resends the whole history with its images every turn).

mrope: `mrope_positions` depends only on the token types and grids BEFORE a row, so identical spans (same start, len, hash -> same grid) give identical positions for every row below `L`; an image that differs cuts `L` at its first row, below which positions are the plain arange shared by both requests. A text-only request after an image request has no spans, so `L` stops at the held image (its later rows used compressed mrope positions the plain rope table would not reproduce). The decode delta is always the new plan's.

Hash: kept at the existing 64-bit SipHash of the encoded bytes (one definition, `vit::image_key`, shared by the tower LRU). Not collision resistant against a crafted image; llama.cpp uses SHA-256 for that reason. serve is a local single-user endpoint and the tower LRU already trusts this key; SHA-256 is a follow-up, not in scope.

Out of scope: the same PNG re-encoded with other bytes is a different image (no reuse, correct but conservative).

Also in this branch (lead's add-on): the boot line `[vit] visual tower loaded: ...` was a `println!` (`gen.rs:1036`, `:1048`), so it reached serve's terminal but not `engine.log` (`grep -c "visual tower" engine.log` = 0 on 2026-09-24). It becomes a tracing event, target `vit`, INFO.

## Expected result

- Unit tests (`cache.rs`): same ids + different image hash -> `L` = image start, reuse only at or below it (cold without such a snapshot); same image -> full reuse including the #100 zero-prefill; text-only -> identical to `decide`; the first differing image wins; an image only one side names is a difference. Each image-difference test red without the fix, green with it.
- Live (robin, about 2 min): restart serve on this build, run `~/Projects/localconf/testcases/diorama-check/preflight.py`: colours **11 of 11**. In `engine.log` each new same-size image shows `[cache] COLD ... images held 1 request 1, image cut L <n> -> <image start>` (or WARM with `P` <= image start), never `prefill 0 of N` for a new image; a re-sent identical image still shows `WARM ... prefill 0`. `grep "visual tower" engine.log` finds the boot line.

**Comment by nibor1896** [comment]
## Implementation

Commit `5bf3fa2` on branch `t-cache-images` (worktree `~/Projects/crow-nest-wt-cacheimg`), local, NOT pushed, not yet merged into `main`. Checked against 0da7420: unchanged since the ticket (same session).

What changed (file:line at 5bf3fa2):
- `engine/src/vit.rs:1298` `image_key` (the one 64-bit hash of the image bytes, now also the tower LRU key), `:1309` `image_spans`, `ImageSpan`; `VisionPlan::spans` filled in `build_plan`.
- `engine/src/cache.rs:232` `common_prefix_len_mm`: id prefix cut to the `start` of the first image span on either side without an identical span on the other (llama.cpp `get_common_prefix`); `:468` `PrefixCache::decide_mm` (`decide_for` = no spans); `:684` `rollback` truncates the held spans.
- `engine/src/gen.rs` `Engine::history_images` + `set_history_images` / `truncate_history_images`; `reset.rs:122` clears it; `slot.rs:668` a slot-file restore clears it.
- `engine/src/bin/serve.rs:3986` `decide_mm` with the plan's spans; `:4046` spans set after the prefill; `[cache]` line appends `images held H request R[, image cut L a -> b]` (text-only line unchanged).
- Add-on: `gen.rs:5059` `log_vit_tower`: the `[vit] visual tower loaded` / `NOT loaded` boot lines are `vit` INFO tracing events (were `println!`).
- `docs/architecture.md` 7.4 "Images (#114)", `cache.rs` module doc, CHANGELOG Unreleased. No new env var; `tools/check_env_docs.py` OK.

Tests (`cargo test --release`, unit only, no GPU job, robin's serve untouched): lib 258 passed / 3 ignored, serve 115 passed, all other bins green.
- Red without the fix (body of `common_prefix_len_mm` reverted to `common_prefix_len(a, b)`, `log_vit_tower` reverted to `println!`), green with it:
  - `cache::tests::same_ids_with_a_different_image_are_not_reused_past_the_image_start` (`L` 1047 != 10)
  - `cache::tests::the_prefix_ends_at_the_first_image_that_differs`
  - `cache::tests::an_image_only_one_side_can_name_is_a_difference`
  - `gen::vit_tower_line::the_tower_line_reaches_the_log_as_a_vit_info_event` (empty writer)
- Guards (green both ways, they pin what must NOT change): `the_same_image_again_is_reused_in_full` (incl. #100 zero-prefill), `text_only_requests_decide_exactly_as_before`, `a_disabled_cache_decides_cold_with_images_too`; `vit::image_identity::*`; `serve` `the_cache_line_names_an_image_cut`.

mrope on reuse: checked in code, not measured. Rows below `L` have identical mrope positions on both sides (positions depend only on types and grids before the row; identical spans imply identical grids); a differing image cuts `L` at its first row; a text-only request after an image request stops at the held image.

NOT verified: no serve run (GPU busy with robin's serve). Before -> after of the colour probe is robin's live check.

Open until robin's live check: rebuild/restart serve on this build, run `~/Projects/localconf/testcases/diorama-check/preflight.py` -> colours 11 of 11; `engine.log` shows `image cut L ... -> <image start>` (COLD or WARM with P below the image) for each new same-size image, never `prefill 0 of N` for a new image; `grep "visual tower" ~/.local/state/crow/logs/engine.log` finds the boot line.

**Comment by nibor1896** [comment]
## Closed on evidence
Before the fix: colours 4 of 11, and `[cache] WARM … prefill 0 of 1047` for new images. After the fix: every new same-size image shows `COLD … image cut L 1029 -> …`, and colours are 11 of 11.
Preflight 2026-09-24 ~12:45 local, serve at crow-nest local main 512ca21 (`~/Projects/localconf/testcases/diorama-check/preflight.py`): boot PASS, colours PASS 11 of 11, render PASS, sees PASS. Closed on robin's instruction to close tickets whose fix is proven.

**Comment by nibor1896** [comment]
## Follow-up: the image identity is SHA-256 (2026-09-24)

- **Commit:** `a34d22b` on branch `t-sampler-hash` (worktree `~/Projects/crow-nest-wt-samphash`, on top of `f0491cd`, #85), local, **not pushed**, not merged.
- **Checked against** `512ca21`: `vit::image_key` (`engine/src/vit.rs:1298` there) was `DefaultHasher::new()` = SipHash-1-3 with fixed keys, 64 bit. It keyed the tower-output cache (`image_cache`), its LRU (`image_lru`) and the prefix cache's `ImageSpan.hash` (`cache::common_prefix_len_mm`).

### Why
Class: GAP / PREVENTION. No collision was observed. A 64-bit collision is a cache HIT for a different image: the other image's tower rows (vit cache) or KV prefix (#114 prefix cache) would answer the request. With fixed, public SipHash keys, a client that picks the bytes could search for such a collision. llama.cpp hashes every mtmd bitmap with SHA-256 for this reason: `tools/mtmd/mtmd-helper.cpp` `mtmd_helper_bitmap_init_from_buf`, lines 385-387 at ggml-org/llama.cpp `3423f94`: `// use sha256 to prevent cache poisoning` / `id = hash_sha256_hex(buf, len);`. `server_tokens::get_common_prefix` compares that id.

### What changed (file:line at `a34d22b`)
- `vit.rs:1294` `pub type ImageKey = [u8; 32]`. `vit.rs:1309` `image_key` = SHA-256 of the encoded bytes. The key stays shared by both caches and the LRU (one definition).
- `vit.rs:1314` `sha256`: in-crate FIPS 180-4 one-shot, about 60 lines. **No new dependency.** `sha2 0.10.9` is in `engine/Cargo.lock` and in the local cargo registry, but only through the optional `cutile-pilot` feature (`cutile-compiler`, `cutile-macro`). Using it would add `sha2` + `digest` + `cpufeatures` + `block-buffer` + `crypto-common` + `generic-array` + `typenum` to the default build. That is robin's call; swapping to `sha2::Sha256::digest` is a one-line change inside `image_key` if he prefers it.
- `ImageSpan.hash`, `image_spans`, `image_cache`, `image_lru`, `cache_insert` take `ImageKey`. The test helper `cache.rs` `img()` spreads its u64 into the 32-byte key.

### Tests
- `vit::image_identity::the_image_key_is_sha256` (`vit.rs:2643`): the NIST example vectors ("abc", "", the 448-bit two-block message, one million 'a'), plus 55/56/64 × 'a' (the padding edges) against coreutils `sha256sum`. **Red without the fix:** with the SipHash key widened to 32 bytes, `left: "c9e61644b3e4a34a0000..."` does not match. Green with the fix.
- `the_image_key_is_the_content_of_the_bytes` (`vit.rs:2620`), extended: same bytes -> same key, one changed byte -> different key, one flipped bit in a 200,003-byte buffer -> different key. This property also holds for the old hash, so this test is **green with and without the fix**. The discriminating test is the SHA-256 one above.
- `cargo test --release`: lib 259 passed / 3 ignored, serve 117, all green. The #114 prefix-cache tests (`cache.rs`) pass on the 32-byte key.

### Not verified
- Hashing cost per image is not measured. It is one pass over the encoded bytes per image per request. Every request hashes every image, including history images Crow re-sends. On a miss, the decode and preprocess of the same bytes also run.
- No serve run was made. robin's live check is the one from #114: colour probe, 11 of 11. `[vit-cache] ... HIT` for a re-sent image shows the key is stable within the process.

**Comment by nibor1896** [comment]
Follow-up ab79523 (local main): `vit::image_key` is now RustCrypto `sha2` (`=0.10.9`, the version Cargo.lock already pinned). The in-crate SHA-256 is removed, on robin's decision: no hand-rolled crypto. The NIST vector test passes. `cargo test --release`: 387 passed.

**Comment by nibor1896** [comment]
Correction to the comment above: the sha2 change is commit **9c9fd51** (local main), not ab79523.
